java/ql/lib/semmle/code/java/frameworks/MyBatis.qll

```diff
@@ -102,3 +102,40 @@ class MyBatisSqlOperationAnnotationMethod extends Method {
 class TypeParam extends Interface {
   TypeParam() { this.hasQualifiedName("org.apache.ibatis.annotations", "Param") }
 }
+
+module ProviderInjection {
+  private import semmle.code.java.dataflow.DataFlow
+
+  class MyBatisInjectionSink extends DataFlow::Node {
+    MyBatisInjectionSink() {
+      exists(Annotation a, Method m, TypeLiteral type, Class c |
+        a.getType()
+            .hasQualifiedName("org.apache.ibatis.annotations",
+              ["Select", "Delete", "Insert", "Update"] + "Provider") and
+        type = a.getValue(["type", "value"]) and
+        c.hasMethod(m, type.getTypeName().getType()) and
+        m.hasName(a.getTarget().getName()) and
+        this.asExpr() = m.getBody().getAStmt().(ReturnStmt).getResult()
+      )
+    }
+  }
+
+  class MyBatisAbstractSQLStep extends Unit {
```

Comment:
I don't think `Abstract` is a good name here, because this class is not abstract. Also, instead of extending `Unit` and having to add this to each relevant query, I'd recommend one of the following:

* Just extend `AdditionalTaintStep`.
* Implement this as rows of `SummaryModelCsv` (argument to return value is a common step for a CSV summary).

Reply:
> I don't think Abstract is a good name here

`Abstract` stands for the name of the class, but I agree this would mislead. How should the class name be then?

> Just extend AdditionalTaintStep.

Done 👍 

> Implement this as rows of SummaryModelCsv (argument to return value is a common step for a CSV summary)

Thank you! I will look into it :)